### PR TITLE
GEODE-8756: Fix CacheableString::objectSize

### PR DESCRIPTION
**Pull request by gaussianrecurrence**
- objectSize was not taking into account null-terminator character,
    which as of C++11, as stated in § 21.4.7.1, should be included at
    the end of the same string character sequence.
  - Also, in those cases in which SSO applied, the returned size was
    higher than the actual one.
  - A parametrized UT was added to check objectSize is returning the
    right size. Note that no fixed sizes are part of such test, given
    that each std::string implementation could have a different SSO
    size.

### CODE REVIEW

File: cppcache/src/CacheableString.cpp [hunk 1]
```diff
@@ -119,8 +119,18 @@ bool CacheableString::isAscii(const std::string& str) {
...
+  auto size = sizeof(CacheableString);
+
+  // This is calculated in order not to count more bytes than necessary
+  // whenever SSO applies.
+  auto delta = reinterpret_cast<const uint8_t*>(m_str.data()) -
+               reinterpret_cast<const uint8_t*>(&m_str);
+  if (delta >= static_cast<decltype(delta)>(sizeof(decltype(m_str))) ||
+      delta < 0) {
+    // Add an extra character for the null-terminator
+    size += sizeof(decltype(m_str)::value_type) * (m_str.capacity() + 1UL);
```
Comment (pivotal-jbarrett) [inline cppcache/src/CacheableString.cpp:131]: Very subtle find on the null terminator!
> Memory locations obtained from the allocator but not available for storing any element are not counted in the allocated storage. Note that the null terminator is not an element of the basic_string.

File: cppcache/test/CacheableStringTests.cpp [hunk 2]
```diff
@@ -224,4 +224,28 @@ TEST_F(CacheableStringTests, TestFromDataNonAsciiHuge) {
...
+                         ::testing::Values(0UL, 1UL, 3UL, 7UL, 15UL, 23UL, 31UL,
+                                           47UL, 63UL, 1025UL, 4097UL));
+
+TEST_P(CacheableStringSizeTests, TestObjectSize) {
+  auto size = GetParam();
+  std::string str(size, '_');
+  auto cacheable = CacheableString::create(str);
+  EXPECT_TRUE(cacheable);
+
+  auto expected_size = sizeof(CacheableString);
```
Comment (pivotal-jbarrett) [inline cppcache/test/CacheableStringTests.cpp:240]: If we are literally copying the internal logic of the method is it really a valid test? Is there another way test the response?

Could we provide a customized allocator that we use for checking allocations? 🤷

**Review by pivotal-jbarrett — APPROVED** [review]
Good find!!

My only concern is the way we are testing this. I am not sure there is really a good way to test it that is platform independent.

File: cppcache/test/CacheableStringTests.cpp (hunk 2 above)
Comment (gaussianrecurrence) [reply cppcache/test/CacheableStringTests.cpp:240]: Actually, it's a good idea!! It should be feasible

Comment (pivotal-jbarrett) [reply cppcache/test/CacheableStringTests.cpp:240]: I think if you use the the custom allocator to allocate a `std::string` in the test itself to get the allocation size you could probably safely assume the same allocations internally. It's possible that an optimization might differ though and internally a different size is allocated. In that case the custom allocator bits have to flow through `CacheableString` that might get tricky.

File: cppcache/test/CacheableStringTests.cpp [hunk 3]
```diff
@@ -224,4 +227,25 @@ TEST_F(CacheableStringTests, TestFromDataNonAsciiHuge) {
...
+};
+
+INSTANTIATE_TEST_SUITE_P(CacheableStringTests, CacheableStringSizeTests,
+                         ::testing::Values(0UL, 1UL, 3UL, 7UL, 15UL, 23UL, 31UL,
+                                           47UL, 63UL, 1025UL, 4097UL));
+
+TEST_P(CacheableStringSizeTests, TestObjectSize) {
+  auto str_size = GetParam();
+  auto expected_size = sizeof(CacheableString);
+  size_tracking_allocator<char> allocator{
```
Comment (pivotal-jbarrett) [inline cppcache/test/CacheableStringTests.cpp:240]: THIS IS AWESOME!!!

File: cppcache/src/util/size_tracking_allocator.hpp [hunk 4]
```diff
@@ -0,0 +1,110 @@
+/*
```
Comment (pivotal-jbarrett) [inline cppcache/src/util/size_tracking_allocator.hpp:1]: This is so cool and probably worth utilizing internally for object size counting in the future. 👏

**Review by pivotal-jbarrett — APPROVED** [review]
This is so very awesome! I would like to see a unit test for the allocator but other than that I good to approve.

**Comment by pdxcodemonkey** [comment]
@gaussianrecurrence This breaks the RHEL7 build, so the change has been reverted in develop, sorry. Please open a new PR when resolved. Here's the relevant output from the RHEL7 build:

In file included from /opt/rh/devtoolset-4/root/usr/include/c++/5.3.1/string:53:0,
from /opt/rh/devtoolset-4/root/usr/include/c++/5.3.1/random:40,
from /opt/rh/devtoolset-4/root/usr/include/c++/5.3.1/bits/stl_algo.h:66,
from /opt/rh/devtoolset-4/root/usr/include/c++/5.3.1/algorithm:62,
from /opt/rh/devtoolset-4/root/usr/include/c++/5.3.1/regex:38,
from /home/geode/nativeclient/cppcache/test/CacheableStringTests.cpp:21:
/opt/rh/devtoolset-4/root/usr/include/c++/5.3.1/bits/basic_string.tcc: In instantiation of ‘static _CharT* std::basic_string<_CharT, _Traits, _Alloc>::_S_construct(std::basic_string<_CharT, _Traits, _Alloc>::size_type, _CharT, const _Alloc&) [with _CharT = char; _Traits = std::char_traits; _Alloc = apache::geode::client::size_tracking_allocator; std::basic_string<_CharT, _Traits, _Alloc>::size_type = long unsigned int]’:
/opt/rh/devtoolset-4/root/usr/include/c++/5.3.1/bits/basic_string.tcc:665:31: required from ‘std::basic_string<_CharT, _Traits, _Alloc>::basic_string(std::basic_string<_CharT, _Traits, _Alloc>::size_type, _CharT, const _Alloc&) [with _CharT = char; _Traits = std::char_traits; _Alloc = apache::geode::client::size_tracking_allocator; std::basic_string<_CharT, _Traits, _Alloc>::size_type = long unsigned int]’
/home/geode/nativeclient/cppcache/test/CacheableStringTests.cpp:243:35: required from here
/opt/rh/devtoolset-4/root/usr/include/c++/5.3.1/bits/basic_string.tcc:600:27: error: no match for ‘operator==’ (operand types are ‘const apache::geode::client::size_tracking_allocator’ and ‘apache::geode::client::size_tracking_allocator’)
if (__n == 0 && __a == _Alloc())
^
In file included from /home/geode/nativeclient/cppcache/test/CacheableStringTests.cpp:33:0:
/home/geode/nativeclient/cppcache/shared/../src/util/size_tracking_allocator.hpp:92:8: note: candidate: bool apache::geode::client::size_tracking_allocator<_Tp>::operator==(const apache::geode::client::size_tracking_allocator<_Tp>&) [with _Tp = char]
bool operator==(size_tracking_allocator const &) { return true; }
^
/home/geode/nativeclient/cppcache/shared/../src/util/size_tracking_allocator.hpp:92:8: note: passing ‘const apache::geode::client::size_tracking_allocator*’ as ‘this’ argument discards qualifiers
In file included from /opt/rh/devtoolset-4/root/usr/include/c++/5.3.1/list:63:0,
from /home/geode/nativeclient/cppcache/shared/../src/PdxType.hpp:23,
from /home/geode/nativeclient/cppcache/shared/../src/PdxTypeRegistry.hpp:33,
from /home/geode/nativeclient/cppcache/shared/../src/CacheImpl.hpp:37,
from /home/geode/nativeclient/cppcache/shared/../src/DataOutputInternal.hpp:26,
from /home/geode/nativeclient/cppcache/test/CacheableStringTests.cpp:30:
/opt/rh/devtoolset-4/root/usr/include/c++/5.3.1/bits/stl_list.h:1807:5: note: candidate: template<class _Tp, class _Alloc> bool std::operator==(const std::list<_Tp, _Alloc>&, const std::list<_Tp, _Alloc>&)
operator==(const list<_Tp, _Alloc>& __x, const list<_Tp, _Alloc>& __y)
^
/opt/rh/devtoolset-4/root/usr/include/c++/5.3.1/bits/stl_list.h:1807:5: note: template argument deduction/substitution failed:
In file included from /opt/rh/devtoolset-4/root/usr/include/c++/5.3.1/string:53:0,
from /opt/rh/devtoolset-4/root/usr/include/c++/5.3.1/random:40,
from /opt/rh/devtoolset-4/root/usr/include/c++/5.3.1/bits/stl_algo.h:66,
from /opt/rh/devtoolset-4/root/usr/include/c++/5.3.1/algorithm:62,
from /opt/rh/devtoolset-4/root/usr/include/c++/5.3.1/regex:38,
from /home/geode/nativeclient/cppcache/test/CacheableStringTests.cpp:21:
/opt/rh/devtoolset-4/root/usr/include/c++/5.3.1/bits/basic_string.tcc:600:27: note: ‘const apache::geode::client::size_tracking_allocator’ is not derived from ‘const std::list<_Tp, _Alloc>’
if (__n == 0 && __a == _Alloc())
^
In file included from /opt/rh/devtoolset-4/root/usr/include/c++/5.3.1/list:63:0,
from /home/geode/nativeclient/cppcache/shared/../src/PdxType.hpp:23,
from /home/geode/nativeclient/cppcache/shared/../src/PdxTypeRegistry.hpp:33,
from /home/geode/nativeclient/cppcache/shared/../src/CacheImpl.hpp:37,
from /home/geode/nativeclient/cppcache/shared/../src/DataOutputInternal.hpp:26,
from /home/geode/nativeclient/cppcache/test/CacheableStringTests.cpp:30:
/opt/rh/devtoolset-4/root/usr/include/c++/5.3.1/bits/stl_list.h:285:5: note: candidate: template bool std::operator==(const std::_List_iterator<_Tp>&, const std::_List_const_iterator<_Val>&)
operator==(const _List_iterator<_Val>& __x,
^
/opt/rh/devtoolset-4/root/usr/include/c++/5.3.1/bits/stl_list.h:285:5: note: template argument deduction/substitution failed:
In file included from /opt/rh/devtoolset-4/root/usr/include/c++/5.3.1/string:53:0,
from /opt/rh/devtoolset-4/root/usr/include/c++/5.3.1/random:40,
from /opt/rh/devtoolset-4/root/usr/include/c++/5.3.1/bits/stl_algo.h:66,
from /opt/rh/devtoolset-4/root/usr/include/c++/5.3.1/algorithm:62,
from /opt/rh/devtoolset-4/root/usr/include/c++/5.3.1/regex:38,
from /home/geode/nativeclient/cppcache/test/CacheableStringTests.cpp:21:
/opt/rh/devtoolset-4/root/usr/include/c++/5.3.1/bits/basic_string.tcc:600:27: note: ‘const apache::geode::client::size_tracking_allocator’ is not derived from ‘const std::_List_iterator<_Tp>’
if (__n == 0 && __a == _Alloc())
^
In file included from /opt/rh/devtoolset-4/root/usr/include/c++/5.3.1/unordered_set:48:0,
from /home/geode/nativeclient/cppcache/include/geode/CacheableKey.hpp:25,
from /home/geode/nativeclient/cppcache/include/geode/CacheableBuiltins.hpp:30,
from /home/geode/nativeclient/cppcache/include/geode/PdxInstanceFactory.hpp:25,
from /home/geode/nativeclient/cppcache/include/geode/RegionService.hpp:26,
from /home/geode/nativeclient/cppcache/include/geode/GeodeCache.hpp:23,
from /home/geode/nativeclient/cppcache/include/geode/Cache.hpp:27,
from /home/geode/nativeclient/cppcache/include/geode/CacheFactory.hpp:25,
from /home/geode/nativeclient/cppcache/test/CacheableStringTests.cpp:25:
/opt/rh/devtoolset-4/root/usr/include/c++/5.3.1/bits/unordered_set.h:1421:5: note: candidate: template<class _Value, class _Hash, class _Pred, class _Alloc> bool std::operator==(const std::unordered_multiset<_Value, _Hash, _Pred, _Alloc>&, const std::unordered_multiset<_Value, _Hash, _Pred, _Alloc>&)
operator==(const unordered_multiset<_Value, _Hash, _Pred, _Alloc>& __x,
^
/opt/rh/devtoolset-4/root/usr/include/c++/5.3.1/bits/unordered_set.h:1421:5: note: template argument deduction/substitution failed:
In file included from /opt/rh/devtoolset-4/root/usr/include/c++/5.3.1/string:53:0,
from /opt/rh/devtoolset-4/root/usr/include/c++/5.3.1/random:40,
from /opt/rh/devtoolset-4/root/usr/include/c++/5.3.1/bits/stl_algo.h:66,
from /opt/rh/devtoolset-4/root/usr/include/c++/5.3.1/algorithm:62,
from /opt/rh/devtoolset-4/root/usr/include/c++/5.3.1/regex:38,
from /home/geode/nativeclient/cppcache/test/CacheableStringTests.cpp:21:
/opt/rh/devtoolset-4/root/usr/include/c++/5.3.1/bits/basic_string.tcc:600:27: note: ‘const apache::geode::client::size_tracking_allocator’ is not derived from ‘const std::unordered_multiset<_Value, _Hash, _Pred, _Alloc>’
if (__n == 0 && __a == _Alloc())
^
In file included from /opt/rh/devtoolset-4/root/usr/include/c++/5.3.1/unordered_set:48:0,
from /home/geode/nativeclient/cppcache/include/geode/CacheableKey.hpp:25,
from /home/geode/nativeclient/cppcache/include/geode/CacheableBuiltins.hpp:30,
from /home/geode/nativeclient/cppcache/include/geode/PdxInstanceFactory.hpp:25,
from /home/geode/nativeclient/cppcache/include/geode/RegionService.hpp:26,
from /home/geode/nativeclient/cppcache/include/geode/GeodeCache.hpp:23,
from /home/geode/nativeclient/cppcache/include/geode/Cache.hpp:27,
from /home/geode/nativeclient/cppcache/include/geode/CacheFactory.hpp:25,
from /home/geode/nativeclient/cppcache/test/CacheableStringTests.cpp:25:
/opt/rh/devtoolset-4/root/usr/include/c++/5.3.1/bits/unordered_set.h:1409:5: note: candidate: template<class _Value, class _Hash, class _Pred, class _Alloc> bool std::operator==(const std::unordered_set<_Value, _Hash, _Pred, _Alloc>&, const std::unordered_set<_Value, _Hash, _Pred, _Alloc>&)
operator==(const unordered_set<_Value, _Hash, _Pred, _Alloc>& __x,
^
/opt/rh/devtoolset-4/root/usr/include/c++/5.3.1/bits/unordered_set.h:1409:5: note: template argument deduction/substitution failed:
In file included from /opt/rh/devtoolset-4/root/usr/include/c++/5.3.1/string:53:0,
from /opt/rh/devtoolset-4/root/usr/include/c++/5.3.1/random:40,
from /opt/rh/devtoolset-4/root/usr/include/c++/5.3.1/bits/stl_algo.h:66,
from /opt/rh/devtoolset-4/root/usr/include/c++/5.3.1/algorithm:62,
from /opt/rh/devtoolset-4/root/usr/include/c++/5.3.1/regex:38,
from /home/geode/nativeclient/cppcache/test/CacheableStringTests.cpp:21:
/opt/rh/devtoolset-4/root/usr/include/c++/5.3.1/bits/basic_string.tcc:600:27: note: ‘const apache::geode::client::size_tracking_allocator’ is not derived from ‘const std::unordered_set<_Value, _Hash, _Pred, _Alloc>’
if (__n == 0 && __a == _Alloc())
^

**Pull request merged**